connector_salesforce/unit/exporter_synchronizer.py

```diff
@@ -0,0 +1,354 @@
+#
+#    Author: Nicolas Bessi
+#    Copyright 2014 Camptocamp SA
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+import logging
+from openerp.addons.connector.queue.job import job
+from openerp.addons.connector.unit.synchronizer import ExportSynchronizer
+from ..unit.rest_api_adapter import with_retry_on_expiration
+
+_logger = logging.getLogger('salesforce_export_synchronizer')
+
+
+class SalesforceExportSynchronizer(ExportSynchronizer):
+    """Exporter to export an Odoo record to Salesforce.
+
+    Exporting using a Salesforce external id field is supported and
+    should  be use only if model is exported only.
+    It is done by setting _sf_lookup in adapter to the name of external field.
+    Switching from export to import direction is possible when using lookup
+    but not the way around. To be able to do this all external fields
+    must be field with SF Object Id.
+    """
+
+    def __init__(self, connector_env):
+        super(SalesforceExportSynchronizer, self).__init__(connector_env)
+        self.salesforce_id = None
+        self.binding_record = None
+        self.binding_id = None
+
+    def _after_export(self, binding_id):
+        """ Hook called after the export"""
+        return
+
+    def _before_export(self):
+        """ Hook called before the export"""
+        return
+
+    def _deactivate(self):
+        """Implementation of the deactivation
+        of an Odoo record on Salesforce
+        """
+        # external id not supported in delete
+        # In Salesforce nothing is deleted it is just a flag
+        # And deleted ind continue to exist into recycle bin.
+        # Recycle bin lifespan is a Salesforce parameter
+        # by default it is 15 days
+        assert self.binding_record.salesforce_id
+        if self.backend_adapter.exists(self.salesforce_id):
+            self.backend_adapter.delete(self.binding_record.salesforce_id)
+
+    def _to_deactivate(self):
+        """Predicate that decide if an Odoo record
+        must be deactivated on Salesforce
+        """
+        assert self.binding_record
+        model = self.session.pool[self.model._name]
+        cols = set(model._columns.keys())
+        cols.update(model._inherit_fields.keys())
```

Comment:
`set(model._all_columns)`? (IIRC the name of the attribute, and `.keys()` is useless)